torch_geometric/nn/conv/hgt_conv.py

```diff
@@ -146,54 +170,58 @@ def forward(
             In case a node type does not receive any message, its output will
             be set to :obj:`None`.
         """
-        H, D = self.heads, self.out_channels // self.heads
+        F = self.out_channels
+        H = self.heads
+        D = F // H
 
         k_dict, q_dict, v_dict, out_dict = {}, {}, {}, {}
 
-        # Iterate over node-types:
-        for node_type, x in x_dict.items():
-            k_dict[node_type] = self.k_lin[node_type](x).view(-1, H, D)
-            q_dict[node_type] = self.q_lin[node_type](x).view(-1, H, D)
-            v_dict[node_type] = self.v_lin[node_type](x).view(-1, H, D)
-            out_dict[node_type] = []
+        # Compute K, Q, V over node types:
+        kqv_dict = self.kqv_lin(x_dict)
+        for key, val in kqv_dict.items():
+            k_dict[key] = val[:, :F].view(-1, H, D)
+            q_dict[key] = val[:, F:2 * F].view(-1, H, D)
+            v_dict[key] = val[:, 2 * F:].view(-1, H, D)
 
-        # Iterate over edge-types:
-        for edge_type, edge_index in edge_index_dict.items():
-            src_type, _, dst_type = edge_type
-            edge_type = '__'.join(edge_type)
+        q, dst_offset = self._cat(q_dict)
+        k, v, src_offset = self._construct_src_node_feat(k_dict, v_dict)
 
-            a_rel = self.a_rel[edge_type]
-            k = (k_dict[src_type].transpose(0, 1) @ a_rel).transpose(1, 0)
+        edge_index, edge_attr = construct_bipartite_edge_index(
+            edge_index_dict, src_offset, dst_offset, edge_attr_dict=self.p_rel)
 
-            m_rel = self.m_rel[edge_type]
-            v = (v_dict[src_type].transpose(0, 1) @ m_rel).transpose(1, 0)
+        out = self.propagate(edge_index, k=k, q=q, v=v, edge_attr=edge_attr,
+                             size=None)
 
-            # propagate_type: (k: Tensor, q: Tensor, v: Tensor, rel: Tensor)
-            out = self.propagate(edge_index, k=k, q=q_dict[dst_type], v=v,
-                                 rel=self.p_rel[edge_type], size=None)
-            out_dict[dst_type].append(out)
+        # Reconstruct output node embeddings dict:
+        for node_type, start_offset in dst_offset.items():
+            end_offset = start_offset + q_dict[node_type].size(0)
+            out_dict[node_type] = out[start_offset:end_offset]
 
-        # Iterate over node-types:
-        for node_type, outs in out_dict.items():
-            out = group(outs, self.group)
+        # Transform output node embeddings:
+        a_dict = self.out_lin({
+            k: torch.nn.functional.gelu(v) if v is not None else v
+            for k, v in out_dict.items()
+        })
 
-            if out is None:
+        # Iterate over node types:
+        for node_type, out in out_dict.items():
+            if out is None or node_type not in self.dst_node_types:
```

Comment:
When would `out is None` be True?

Reply:
I ran hgt and fast hgt for the hgt_dblp.py example and get matching results:
[hgt_vs_fasthgt.txt](https://github.com/pyg-team/pytorch_geometric/files/11187738/hgt_vs_fasthgt.txt)
also checked that their outputs matched w/ matching params
https://github.com/puririshi98/rgcn_pyg_lib_forward_bench/blob/main/hgt_correctness.py